### PR TITLE
Separating Battle logic from Program.lua

### PR DESCRIPTION
**Pull request by UTDZac**
Initial effort of separating out a ton of battle logic from `Program.lua` into a new file, `Battle.lua`. Wanting to get this merged in sooner rather than later (and update `bugfixes` branch afterwards), so others can build off of it.

The main goal is separating Program/Battle based on the responsibilities of each: 
- **Program** likely handles anything related to the frame loops, the timings for handling input and drawing. It also does everything that involves reading data from memory (a few loose ends in Drawing/Utils). 
- **Battle** manages some of the Programs data at a more granular level, and tracks any entire fight. This allows better data management for when in a battle, more accurate data reads, and condensing some of the redundant code from a few of our recent features. (granted, not many improvements right now, but its a start)

Eventually we can have `Battle.lua` provide us with a simplified set of data to draw to the screen. Currently all of this logic is scattered throughout `TrackerScreen.drawMovesArea(...)` which has to perform all of the "calculations" inside a "drawing" function. These ideally would be separated.

Also worth noting, we should probably test this a good bit, especially with some of the ability triggers, and very especially synchronize.

### CODE REVIEW

File: ironmon_tracker/data/MoveData.lua [hunk 1]
```diff
@@ -1,4 +1,6 @@
-MoveData = {}
+MoveData = {
+	totalMoves = 354,
```
Comment (Fellshadow) [inline ironmon_tracker/data/MoveData.lua:2]: any particular reason to move the length to a hardcoded variable instead of using # to get the length?

I don't forsee us adding/removing moves/pokemon to be fair so it's not an issue, but if we were to then would also need to remember to update the totalMoves/totalPokemon variables. I'm just curious as to why you felt this change was needed

**Review by Fellshadow — COMMENTED** [review]
Just a couple things from looking at the code that i've mentioned but other than that i'm happy for it to go in, though i'm gonna do some testing first of at least some ability stuff and what-not. Will update here how that goes

Also there was a big breaking bug introduced in recent bugfixes, i've put through a fix on bugfixes/staging so may wanna update this branch with it too (though there's no conflict here from it so it's probably fine regardless, just noting it)

**Comment by Fellshadow** [comment]
Update, did some brief ability testing and it seemed to work fine. Don't have the means to test a lot of edge-cases like synchronize going off from your static but brief synchronize testing from proccing with moves worked fine

File: ironmon_tracker/data/MoveData.lua (hunk 1 above)
Comment (UTDZac) [reply ironmon_tracker/data/MoveData.lua:2]: I want to say I opted to do this because using `#` to get the size of the table might perform a calculation that has to determine how big the table is, then return that value; as opposed to us just setting the value here and using it. Maybe with Lua that is not the case.

File: ironmon_tracker/screens/TrackerScreen.lua [hunk 2]
```diff
@@ -632,16 +634,15 @@ function TrackerScreen.drawPokemonInfoArea(pokemon)
 			Drawing.drawButton(TrackerScreen.Buttons.PCHealAutoTracking, shadowcolor)
 		end
 	else
-		if Tracker.Data.trainerID ~= nil and pokemon.trainerID ~= nil then
+		-- Assumes we are in Battle, unsure if this check is neccessary
```
[the remark this answers is not in the document]
Comment (UTDZac) [reply ironmon_tracker/screens/TrackerScreen.lua:637]: Good catch and smart habit to avoid assuming anything. I suppose what I was thinking here was eventually we'd extend the route info stuff to work outside of battle, mostly for mapping trainers.

**Review by Fellshadow — APPROVED** [review]
ye looks good to me
